test/fail_compilation/retscope6.d

```diff
@@ -21,3 +21,55 @@ int* test() @safe
     arr[0] ~= &i;
     return arr[0][0];
 }
+
+/* TEST_OUTPUT:
+---
+fail_compilation/retscope6.d(7034): Error: reference to local variable `i` assigned to non-scope parameter `_param_1` calling retscope6.S.emplace!(int*).emplace
+fail_compilation/retscope6.d(7035): Error: reference to local variable `i` assigned to non-scope parameter `_param_0` calling retscope6.S.emplace2!(int*).emplace2
+fail_compilation/retscope6.d(7024): Error: scope variable `_param_2` assigned to `s` with longer lifetime
+fail_compilation/retscope6.d(7025): Error: scope variable `_param_2` assigned to `t` with longer lifetime
+fail_compilation/retscope6.d(7037): Error: template instance `retscope6.S.emplace4!(int*)` error instantiating
+---
+*/
+
+#line 7000
+
+alias T = int*;
+
+struct S
+{
+    T payload;
+
+    static void emplace(Args...)(ref S s, Args args) @safe
+    {
+        s.payload = args[0];
+    }
+
+    void emplace2(Args...)(Args args) @safe
+    {
+        payload = args[0];
+    }
+
+    static void emplace3(Args...)(S s, Args args) @safe
+    {
+        s.payload = args[0];
+    }
+
+    static void emplace4(Args...)(scope ref S s, scope out S t, scope Args args) @safe
+    {
+        s.payload = args[0];
+        t.payload = args[0];
```

Comment:
Due to the lack of documentation surrounding the semantics of `scope` I'm assuming that in this context, `scope` means that `arg[0]`'s lifetime must not exceed the lifetime of this `emplace4`'s scope.  However, since we are assigning to a `ref` and `out` we are escaping out of `emplace4`'s scope.  Is this correct?

When I analyze the implementation of `foo`, however, there's not escape because everything is local to `foo`.  Perhaps this could use a more compelling test that would actually result in a leak or memory corruption without this PR.

Reply:
I'll add some more documentation.

Reply:
@WalterBright Is it really too much to ask to have you open a dlang.org PR to document the semantics of `scope` and `return` function parameters with illustrative examples so we can refer to it when we review these?  Or are you still trying to figure all this out, and this is your way of working through it?

Reply:
It's documented in DIP25 and DIP1000. I'd start with a gentler introduction, http://dconf.org/2017/talks/bright.html